test_quality_of_service/test/test_deadline.cpp

```diff
@@ -0,0 +1,121 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <memory>
+#include <string>
+#include <tuple>
+
+#include "gtest/gtest.h"
+
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp/executors/single_threaded_executor.hpp"
+
+#include "std_msgs/msg/string.hpp"
+
+#include "test_quality_of_service/qos_utilities.hpp"
+
+
+using namespace std::chrono_literals;
+
+/// Test Deadline with a single publishing node and single subscriber node
+TEST_F(TestSetup, test_deadline) {
```

Comment:
Why not get rid of all the locks and do:

```c++
TEST_F(...) {
  // init
  start_publishing();
  spin for 10 sec;
  stop_publishing();
  spin for 10 sec;
  check deadline
}
```
sounds way simplier than making this MT?

Also, I *believe* that sending messages without spinning is an easy way to miss deadlines:

```c++
set deadline to 1 second
send message
wait 5 seconds
spin until message is received somehow
check that we actually missed the deadline
```

Reply:
I didn't have that experience when attempting to write these tests :-( and found it easier to use timers as implemented.